### PR TITLE
More symbol info about collisions in error message when a link is ambiguous

### PR DESCRIPTION
**Pull request by d-ronnqvist**
Bug/issue #, if applicable: https://github.com/apple/swift-docc/issues/397 rdar://100436859

## Summary

This shows the full symbol declarations in the error message for a symbol link that's ambiguous to help developers distinguish between the colliding symbols.

This also fixes a minor issue where the descriptions of unresolved references contained percent encoding, resulting in hard to read diagnostics. 

These changes together refines this error message
```
Topic reference 'emit(_%3A)' couldn't be resolved. Reference is ambiguous after '/SwiftDocC/DiagnosticEngine': Add '1xqdb' to refer to 'emit(_:)'. Add '5rcke' to refer to 'emit(_:)'. 
```
into this error message
```
Topic reference 'emit(_:)' couldn't be resolved. Reference is ambiguous after '/SwiftDocC/DiagnosticEngine': Append '-1xqdb' to refer to 'func emit(_ problem: Problem)'. Append '-5rcke' to refer to 'func emit(_ problems: [Problem])'. 
```

## Dependencies

n/a

## Testing

With `DOCC_USE_HIERARCHY_BASED_LINK_RESOLVER=YES` write an ambiguous link, for example by adding this code
```swift 
public enum CollisionsWithDifferentFunctionArguments {
    public func something(argument: Int) -> Int { 0 }
    public func something(argument: String) -> Int { 0 }
}
```
and linking to `CollisionsWithDifferentFunctionArguments/something(argument:)`

In the resulting error message it should be possible to distinguish the collisions. 

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [x] ~Updated documentation if necessary~

### CODE REVIEW

File: Tests/SwiftDocCTests/Infrastructure/PathHierarchyTests.swift [hunk 1]
```diff
@@ -295,6 +294,11 @@ class PathHierarchyTests: XCTestCase {
             (symbolID: "s:14MixedFramework28CollisionsWithDifferentKindsO9somethingyA2CmF", disambiguation: "enum.case"),
             (symbolID: "s:14MixedFramework28CollisionsWithDifferentKindsO9somethingSSvp", disambiguation: "property"),
         ])
+        try assertPathRaisesErrorMessage("/MixedFramework/CollisionsWithDifferentKinds/something", in: tree, context: context, expectedErrorMessage: """
+        Reference is ambiguous after '/MixedFramework/CollisionsWithDifferentKinds': \
+        Add 'enum.case' to refer to 'case something'. \
```
Comment (franklinsch) [inline Tests/SwiftDocCTests/Infrastructure/PathHierarchyTests.swift:299]: Two suggestions:
- Should we use 'Append' rather than 'Add' so that it's clear where to add the suffix?
- Should we include the hyphen, as in `'-enum.case'` so that it's clearer how to append the suffix?

Comment (d-ronnqvist) [reply Tests/SwiftDocCTests/Infrastructure/PathHierarchyTests.swift:299]: Yes, I think both of those would be good clarifications to the error message.

Comment (d-ronnqvist) [reply Tests/SwiftDocCTests/Infrastructure/PathHierarchyTests.swift]: Resolved in https://github.com/apple/swift-docc/pull/398/commits/0d3557fb4090dfcd8aed27d16604d72d78340374

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift [hunk 2]
```diff
@@ -838,7 +838,7 @@ extension PathHierarchy.Error {
             return "Symbol links can only resolve symbols."
             
         case .lookupCollision(let partialResult, let collisions):
-            let collisionDescription = collisions.map { "Add \($0.disambiguation.singleQuoted) to refer to \($0.node.fullNameOfValue(context: context).singleQuoted)"}.sorted()
+            let collisionDescription = collisions.map { "Append '-\($0.disambiguation)' to refer to \($0.node.fullNameOfValue(context: context).singleQuoted)"}.sorted()
```
Comment (Kyle-Ye) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift:841]: Should we also delete the following "singleQuoted" call？

```suggestion
            let collisionDescription = collisions.map { "Append '-\($0.disambiguation)' to refer to \($0.node.fullNameOfValue(context: context))"}.sorted()
```

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift [hunk 3]
```diff
@@ -864,6 +864,9 @@ private extension PathHierarchy.Node {
     func fullNameOfValue(context: DocumentationContext) -> String {
         guard let identifier = identifier else { return name }
         if let symbol = symbol {
+            if let fragments = symbol[mixin: SymbolGraph.Symbol.DeclarationFragments.self]?.declarationFragments {
+                return fragments.map { $0.spelling }.joined().split(whereSeparator: { $0.isWhitespace || $0.isNewline }).joined(separator: " ")
```
Comment (Kyle-Ye) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift:868]: A small suggestion: Could we remove `whereSeparator` label? It seems a litter inconsistant.
```suggestion
                return fragments.map { $0.spelling }.joined().split { $0.isWhitespace || $0.isNewline }.joined(separator: " ")
```

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift [hunk 4]
```diff
@@ -864,6 +864,9 @@ private extension PathHierarchy.Node {
...
+            }
             return context.symbolIndex[symbol.identifier.precise]!.name.description
```
Comment (Kyle-Ye) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift:870]: A small suggestion. But maybe we should prefer to use
```swift
if xx { 
    return A
} else {
    return B
}
``` 
than
```swift
if xx { 
    return A
}
return B
```

```suggestion
else {
    return context.symbolIndex[symbol.identifier.precise]!.name.description
}        
```

**Review by Kyle-Ye — APPROVED** [review]

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift (hunk 4 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift:870]: We seem to be doing a bit of both styles throughout the code base. I don't have a strong preference for either style. When both the A and B are one line expressions (like in this case) I find the structure of the code easily to skim based on the leading `return` keywords so close to each other.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift (hunk 2 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift:841]: I don't think so. I feel the the quotation marks makes it easier to follow where each declaration starts and ends in the error message. 

I surrounded the disambiguation interpolation with quotation marks because I also needed to prefix it with a dash and I felt that `'-\($0.disambiguation)'` was more readable than `\(("-" + $0.disambiguation).singleQuoted)`. In the rest of the code I think that we're mostly using `.singleQuoted`.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift (hunk 3 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift:868]: I personally find that the argument label for `split` makes it easier for me to contextualize the closure when reading it. I don't mind the surrounding parenthesis being inconsistent but if I were to make it consistent my preference would be 
```suggestion
                return fragments.map({ $0.spelling }).joined().split(whereSeparator: { $0.isWhitespace || $0.isNewline }).joined(separator: " ")
```

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift (hunk 4 above)
Comment (Kyle-Ye) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift:870]: Got it.

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift (hunk 3 above)
Comment (daniel-grumberg) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift:868]: I agree about the argument label `fragments.map(\.spelling).joined().split(whereSeparator: { $0.isWhitespace || $0.isNewline }).joined(separator: " ")`

File: Tests/SwiftDocCTests/Infrastructure/PathHierarchyTests.swift [hunk 5]
```diff
@@ -75,8 +75,7 @@ class PathHierarchyTests: XCTestCase {
         try assertFindsPath("/MixedFramework/MyObjectiveCCompatibleProtocol/myProtocolMethod", in: tree, asSymbolID: "c:@M@MixedFramework@objc(pl)MyObjectiveCCompatibleProtocol(im)myProtocolMethod")
         try assertFindsPath("/MixedFramework/MyObjectiveCCompatibleProtocol/myProtocolMethod()", in: tree, asSymbolID: "c:@M@MixedFramework@objc(pl)MyObjectiveCCompatibleProtocol(im)myProtocolMethod")
         try assertFindsPath("/MixedFramework/MyObjectiveCCompatibleProtocol/myProtocolProperty", in: tree, asSymbolID: "c:@M@MixedFramework@objc(pl)MyObjectiveCCompatibleProtocol(py)myProtocolProperty")
-        // Objective-C class properties have a "property" kind instead of a "type.property" kind (rdar://92927788)
```
Comment (daniel-grumberg) [inline Tests/SwiftDocCTests/Infrastructure/PathHierarchyTests.swift:78]: This seems unrelated, might be worth splitting this into a separate commit.

**Review by daniel-grumberg — APPROVED** [review]

File: Tests/SwiftDocCTests/Infrastructure/PathHierarchyTests.swift (hunk 5 above)
Comment (d-ronnqvist) [reply Tests/SwiftDocCTests/Infrastructure/PathHierarchyTests.swift:78]: It is a separate commit in the PR. Did you mean open a separate PR for this change?

**Comment by d-ronnqvist** [comment]
@swift-ci please test
